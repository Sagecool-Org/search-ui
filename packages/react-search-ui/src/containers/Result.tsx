import React from "react";
import { Component } from "react";
import {
  Result,
  ResultContainerProps,
  ResultViewProps
} from "@elastic/react-search-ui-views";
import { useSearch } from "../hooks";

export const ResultContainer = ({
  result,
  shouldTrackClickThrough,
  clickThroughTags,
  view,
  ...rest
}: ResultContainerProps) => {
  const {
    driver: {
      actions: { trackClickThrough }
    }
  } = useSearch();

  const handleClickLink = (id: string) => {
    if (shouldTrackClickThrough) {
      trackClickThrough(id, clickThroughTags);
    }
  };
  const View = view || Result;
  const id = result.id.raw;
  const viewProps: ResultViewProps = {
    result: result,
    key: `result-${id}`,
    onClickLink: () => handleClickLink(id),
    ...rest
  };

  return <View {...viewProps} />;
};

export default ResultContainer;
