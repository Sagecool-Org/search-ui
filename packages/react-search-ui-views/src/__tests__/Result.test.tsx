import React from "react";
import { render, fireEvent } from "@testing-library/react";
import Result from "../Result";

const TITLE_FIELD = "title";
const URL_FIELD = "url";
const TITLE_RESULT_VALUE = "Title";
const URL_RESULT_VALUE = "http://www.example.com";
const ARBITRARY_FIELD = { _meta: "data" };

const requiredProps = {
  result: { field: { raw: "value" } },
  onClickLink: jest.fn()
};

it("renders correctly when there is not a URL or title", () => {
  const { container } = render(<Result {...requiredProps} />);
  expect(container).toMatchSnapshot();
});

it("renders correctly when there is a title", () => {
  const { container } = render(
    <Result
      {...{
        ...requiredProps,
        result: {
          ...requiredProps.result,
          [TITLE_FIELD]: { raw: TITLE_RESULT_VALUE }
        },
        titleField: TITLE_FIELD
      }}
    />
  );
  expect(container).toMatchSnapshot();
  expect(container.querySelector(".sui-result__title")).toHaveTextContent(
    TITLE_RESULT_VALUE
  );
});

it("renders correctly when there is a titleField but it is not defined in result", () => {
  const { container } = render(
    <Result {...{ ...requiredProps, titleField: TITLE_FIELD }} />
  );
  expect(container).toMatchSnapshot();
  expect(container.querySelector(".sui-result__title")).not.toBeInTheDocument();
});

it("renders correctly when there is a URL", () => {
  const { container } = render(
    <Result
      {...{
        ...requiredProps,
        result: {
          ...requiredProps.result,
          [URL_FIELD]: { raw: URL_RESULT_VALUE }
        },
        urlField: URL_FIELD
      }}
    />
  );
  expect(container).toMatchSnapshot();
  const link = container.querySelector(".sui-result__title-link");
  expect(link).toHaveAttribute("href", URL_RESULT_VALUE);
});

it("renders correctly when there is a urlField but it is not defined in result", () => {
  const { container } = render(
    <Result {...{ ...requiredProps, urlField: URL_FIELD }} />
  );
  expect(container).toMatchSnapshot();
  expect(
    container.querySelector(".sui-result__title-link")
  ).not.toBeInTheDocument();
});

it("renders correctly when there is a title and url", () => {
  const { container } = render(
    <Result
      {...{
        ...requiredProps,
        result: {
          ...requiredProps.result,
          [TITLE_FIELD]: { raw: TITLE_RESULT_VALUE },
          [URL_FIELD]: { raw: URL_RESULT_VALUE }
        },
        titleField: TITLE_FIELD,
        urlField: URL_FIELD
      }}
    />
  );
  expect(container).toMatchSnapshot();
  const link = container.querySelector(".sui-result__title-link");
  expect(link).toHaveAttribute("href", URL_RESULT_VALUE);
  expect(link).toHaveTextContent(TITLE_RESULT_VALUE);
});

it("filters out arbitrary fields from results, and does not render them", () => {
  const { container } = render(
    <Result
      {...{
        ...requiredProps,
        result: {
          ...requiredProps.result,
          ...ARBITRARY_FIELD
        }
      }}
    />
  );
  expect(container).toMatchSnapshot();
  expect(container.querySelector(".sui-result__key")).not.toHaveTextContent(
    "_meta"
  );
});

it("renders with className prop applied", () => {
  const customClassName = "test-class";
  const { container } = render(
    <Result className={customClassName} {...requiredProps} />
  );
  expect(container.firstChild).toHaveClass("sui-result", "test-class");
});

it("renders correctly when there is a malicious URL", () => {
  const { container } = render(
    <Result
      {...{
        ...requiredProps,
        result: {
          ...requiredProps.result,
          [URL_FIELD]: { raw: "javascript://test%0aalert(document.domain)" }
        },
        urlField: URL_FIELD
      }}
    />
  );
  expect(container).toMatchSnapshot();
  expect(
    container.querySelector(".sui-result__title-link")
  ).not.toBeInTheDocument();
});

it("calls onClickLink when clicking on the link", () => {
  const { container } = render(
    <Result
      {...{
        ...requiredProps,
        result: {
          ...requiredProps.result,
          [TITLE_FIELD]: { raw: TITLE_RESULT_VALUE },
          [URL_FIELD]: { raw: URL_RESULT_VALUE }
        },
        titleField: TITLE_FIELD,
        urlField: URL_FIELD
      }}
    />
  );
  const link = container.querySelector(".sui-result__title-link");
  fireEvent.click(link);
  expect(requiredProps.onClickLink).toHaveBeenCalled();
});
