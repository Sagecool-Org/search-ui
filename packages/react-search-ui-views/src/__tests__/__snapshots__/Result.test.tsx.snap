// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`filters out arbitrary fields from results, and does not render them 1`] = `
<div>
  <li
    class="sui-result"
  >
    <div
      class="sui-result__header"
    />
    <div
      class="sui-result__body"
    >
      <ul
        class="sui-result__details"
      >
        <li>
          <span
            class="sui-result__key"
          >
            field
          </span>
           
          <span
            class="sui-result__value"
          >
            value
          </span>
        </li>
      </ul>
    </div>
  </li>
</div>
`;

exports[`renders correctly when there is a URL 1`] = `
<li
  className="sui-result"
>
  <div
    className="sui-result__header"
  />
  <div
    className="sui-result__body"
  >
    <ul
      className="sui-result__details"
    >
      <li
        key="field"
      >
        <span
          className="sui-result__key"
        >
          field
        </span>
         
        <span
          className="sui-result__value"
          dangerouslySetInnerHTML={
            Object {
              "__html": "value",
            }
          }
        />
      </li>
      <li
        key="url"
      >
        <span
          className="sui-result__key"
        >
          url
        </span>
         
        <span
          className="sui-result__value"
          dangerouslySetInnerHTML={
            Object {
              "__html": "http://www.example.com",
            }
          }
        />
      </li>
    </ul>
  </div>
</li>
`;

exports[`renders correctly when there is a malicious URL 1`] = `
<div>
  <li
    class="sui-result"
  >
    <div
      class="sui-result__header"
    />
    <div
      class="sui-result__body"
    >
      <ul
        class="sui-result__details"
      >
        <li>
          <span
            class="sui-result__key"
          >
            field
          </span>
           
          <span
            class="sui-result__value"
          >
            value
          </span>
        </li>
        <li>
          <span
            class="sui-result__key"
          >
            url
          </span>
           
          <span
            class="sui-result__value"
          >
            javascript://test%0aalert(document.domain)
          </span>
        </li>
      </ul>
    </div>
  </li>
</div>
`;

exports[`renders correctly when there is a title 1`] = `
<div>
  <li
    class="sui-result"
  >
    <div
      class="sui-result__header"
    >
      <span
        class="sui-result__title"
      >
        Title
      </span>
    </div>
    <div
      class="sui-result__body"
    >
      <ul
        class="sui-result__details"
      >
        <li>
          <span
            class="sui-result__key"
          >
            field
          </span>
           
          <span
            class="sui-result__value"
          >
            value
          </span>
        </li>
        <li>
          <span
            class="sui-result__key"
          >
            title
          </span>
           
          <span
            class="sui-result__value"
          >
            Title
          </span>
        </li>
      </ul>
    </div>
  </li>
</div>
`;

exports[`renders correctly when there is a title and url 1`] = `
<div>
  <li
    class="sui-result"
  >
    <div
      class="sui-result__header"
    >
      <a
        class="sui-result__title sui-result__title-link"
        href="http://www.example.com"
        rel="noopener noreferrer"
        target="_blank"
      >
        Title
      </a>
    </div>
    <div
      class="sui-result__body"
    >
      <ul
        class="sui-result__details"
      >
        <li>
          <span
            class="sui-result__key"
          >
            field
          </span>
           
          <span
            class="sui-result__value"
          >
            value
          </span>
        </li>
        <li>
          <span
            class="sui-result__key"
          >
            title
          </span>
           
          <span
            class="sui-result__value"
          >
            Title
          </span>
        </li>
        <li>
          <span
            class="sui-result__key"
          >
            url
          </span>
           
          <span
            class="sui-result__value"
          >
            http://www.example.com
          </span>
        </li>
      </ul>
    </div>
  </li>
</div>
`;

exports[`renders correctly when there is a titleField but it is not defined in result 1`] = `
<div>
  <li
    class="sui-result"
  >
    <div
      class="sui-result__header"
    />
    <div
      class="sui-result__body"
    >
      <ul
        class="sui-result__details"
      >
        <li>
          <span
            class="sui-result__key"
          >
            field
          </span>
           
          <span
            class="sui-result__value"
          >
            value
          </span>
        </li>
      </ul>
    </div>
  </li>
</div>
`;

exports[`renders correctly when there is a urlField but it is not defined in result 1`] = `
<div>
  <li
    class="sui-result"
  >
    <div
      class="sui-result__header"
    />
    <div
      class="sui-result__body"
    >
      <ul
        class="sui-result__details"
      >
        <li>
          <span
            class="sui-result__key"
          >
            field
          </span>
           
          <span
            class="sui-result__value"
          >
            value
          </span>
        </li>
      </ul>
    </div>
  </li>
</div>
`;

exports[`renders correctly when there is not a URL or title 1`] = `
<div>
  <li
    class="sui-result"
  >
    <div
      class="sui-result__header"
    />
    <div
      class="sui-result__body"
    >
      <ul
        class="sui-result__details"
      >
        <li>
          <span
            class="sui-result__key"
          >
            field
          </span>
           
          <span
            class="sui-result__value"
          >
            value
          </span>
        </li>
      </ul>
    </div>
  </li>
</div>
`;
